mips_api/mips.py

```diff
@@ -111,12 +119,76 @@ def _collect_mips_data(self):
                 headers={"Authorization-Token":access_token},
             )
 
+    def _mips_dict_json(self):
+        '''
+        Return the full chart of accounts as JSON
+        '''
+        return json.dumps(self.mips_dict, indent=2)
+
+    def _tag_list_json(self):
+        '''
+        Generate a list of valid AWS tags. Only active codes are listed.
+
+        The string format is `{Program Name} / {Program Code}`.
+
+        Returns
+            A JSON string representing an array of strings.
+        '''
+
+        data = []
+
+        for code, name in self._extra_program_codes.items():
+            data.append(f"{name} / {code}")
+
+        # inactive codes have 5 digits, active codes have 8;
+        # and only the first 6 digits of active codes are significant
+        for code, name in self.mips_dict.items():
+            if len(code) > 5: # only include active codes
+                short = code[:6]  # ignore the last two digits on active codes
+                if short not in self._omit_program_codes:
+                    title = f"{name} / {short}"
+                    if title not in data:
+                        data.append(title)
+
+        return json.dumps(data, indent=2)
+
+    def valid_routes(self):
+        '''
+        List all valid routes
+        '''
+        return self.api_routes.values()
+
+    def _get_mips_data(self, lookup):
+        '''
+        Process MIPS data into requested format
+        '''
+
+        if lookup == self.api_routes['apiChartOfAccounts']:
+            return self._mips_dict_json()
+
+        if lookup == self.api_routes['apiValidTags']:
+            return self._tag_list_json()
+
+        raise Exception('Invalid API route')
+
     def get_mips_data(self, lookup):
-        '''TODO: implement cache lookup'''
-        if lookup not in self.valid_routes:
-            raise Exception(f"Invalid request: {lookup}")
+        '''
+        Entry point for retrieving data
 
+        Returns
+            The body of the requested object
+        '''
+
+        # Verify a valid lookup path
+        if lookup not in self.valid_routes():
+            raise Exception(f'Invalid API route: {lookup}')
+
+        # Query MIPS if needed
```

Comment:
The lambda is not stateless, it keeps both ssm secrets and the third-party response in memory, and it needs to serialize access to the third-party api. also if the same lambda environment is reused (i.e. the lambda is run within 15 minutes of a previous run) then it becomes re-entrant. so while most lambda runs are expected to have a cold start, we also need to support both warm starts and running in parallel (and ensuring only one run logs in to the third-party api at a time). 

the approach here is to be stateful so that warm starts (either in parallel or not) rely on state from the cold start, avoiding any secondary attempts to access the third-party api.

Reply:
hmm, i'm our understand matches with how re-entrant lambdas work.  i don't think stuff in memory persists for the re-entrant lambda use case.

Reply:
hmm, i'm not sure that our understand matches with how re-entrant lambdas work. i don't think stuff in memory persists for the re-entrant lambda use case.